compiler/rustc_mir_transform/src/match_branches.rs

```diff
@@ -515,3 +525,210 @@ impl<'tcx> SimplifyMatch<'tcx> for SimplifyToExp {
         }
     }
 }
+
+/// This is primarily used to merge these copy statements that simplified the canonical enum clone method by GVN.
+/// The GVN simplified
+/// ```ignore (syntax-highlighting-only)
+/// match a {
+///     Foo::A(x) => Foo::A(*x),
+///     Foo::B => Foo::B
+/// }
+/// ```
+/// to
+/// ```ignore (syntax-highlighting-only)
+/// match a {
+///     Foo::A(_x) => a, // copy a
+///     Foo::B => Foo::B
+/// }
+/// ```
+/// This function will simplify into a copy statement.
+fn simplify_to_copy<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &mut Body<'tcx>,
+    switch_bb_idx: BasicBlock,
+    param_env: ParamEnv<'tcx>,
+) -> Option<()> {
+    if switch_bb_idx != START_BLOCK {
+        return None;
+    }
+    let bbs = &body.basic_blocks;
+    // Check if the copy source matches the following pattern.
+    // _2 = discriminant(*_1); // "*_1" is the expected the copy source.
+    // switchInt(move _2) -> [0: bb3, 1: bb2, otherwise: bb1];
+    let &Statement {
+        kind: StatementKind::Assign(box (discr_place, Rvalue::Discriminant(expected_src_place))),
+        ..
+    } = bbs[switch_bb_idx].statements.last()?
+    else {
+        return None;
+    };
+    let expected_src_ty = expected_src_place.ty(body.local_decls(), tcx);
+    if !expected_src_ty.ty.is_enum() || expected_src_ty.variant_index.is_some() {
+        return None;
+    }
+    let expected_dest_place = Place::return_place();
+    let expected_dest_ty = expected_dest_place.ty(body.local_decls(), tcx);
+    if expected_dest_ty.ty != expected_src_ty.ty || expected_dest_ty.variant_index.is_some() {
+        return None;
+    }
+    let targets = match bbs[switch_bb_idx].terminator().kind {
+        TerminatorKind::SwitchInt { ref discr, ref targets, .. }
+            if discr.place() == Some(discr_place) =>
+        {
+            targets
+        }
+        _ => return None,
+    };
+    // We require that the possible target blocks all be distinct.
+    if !targets.is_distinct() {
+        return None;
+    }
+    if !bbs[targets.otherwise()].is_empty_unreachable() {
+        return None;
+    }
+    // Check that destinations are identical, and if not, then don't optimize this block.
+    let mut target_iter = targets.iter();
+    let first_terminator_kind = &bbs[target_iter.next().unwrap().1].terminator().kind;
+    if !target_iter
+        .all(|(_, other_target)| first_terminator_kind == &bbs[other_target].terminator().kind)
+    {
+        return None;
+    }
+
+    let borrowed_locals = borrowed_locals(body);
```

Comment:
This visits the whole body, so this should only be computed once and passed as a parameter to this function. Otherwise, you get quadratic in the number of bbs.

Reply:
> This visits the whole body, so this should only be computed once and passed as a parameter to this function. 

I checked if the BB is `START_BLOCK`.

> Otherwise, you get quadratic in the number of bbs.

Do you mean the compile-time?